core/iwasm/common/wasm_c_api.c

```diff
@@ -2004,13 +2014,33 @@ wasm_module_delete_internal(wasm_module_t *module)
     }
 
     module_ex = module_to_module_ext(module);
+
+    /* deleting the module... */
+    if (module_ex->ref_count == 0) {
```

Comment:
Should move the check after locking the `module_ex->lock`? Suppose here the module_ex->ref_count read is 1 and we continue to run. And another thread decreases the count to 0 before this thread runs into lock L2024, then in L2028, after decreasing in L2027, the count is now -1 (or UINT32_MAX), the code will continue to run into L2033 to 2044, which is unexpected behavior?

Reply:
I thought there are cases when hitting L2024 : `os_mutex_lock(&module_ex->lock);`
- `module_ex` has been released. Dereference it is UB. It depends on callers' behavior to avoid it.
- `module_ex->lock` has been destroyed. It leads a failed return code of `os_muext_lock`.  Guess it is the "unexpected behavior` mentioned above?
- `module_ex->ref_count` is zero. In this case `ref_count--` will be `UINT32_MAX` and `if >0` will let `return` work
- `module_ex->ref_count` is great than zero.